### DISCUSSION TITLE
How to embed content from Jupyter Notebook into a separated Markdown file?

### DISCUSSION THREAD

**Discussion by lgonzalezsa** [executablebooks/meta#1062] [Q&A] [answered]
Exploring Myst Markdown and found [https://mystmd.org/guide/reuse-jupyter-outputs](https://mystmd.org/guide/reuse-jupyter-outputs) which is something I would like to do with my data analysis.
1. Do the data collection and exploration in a Jupyter Notebook (using vscode by the way)
2. Create an isolated Markdown file to write my article, conclusions, summary of results but embedding the cell output of the other Jupyter Notebooks created for the solution

I was exploring the documentation but in addition to the link above about reuse jupyter outputs, did not find an example to follow.

I was also exploring Quarto [https://quarto.org/docs/authoring/notebook-embed#code-cell-options](https://quarto.org/docs/authoring/notebook-embed#code-cell-options) which seems to support this method, however, I would like to try Myst.

Any pointers to examples I really appreciate.

**Comment by fwkoch** [comment, marked as answer]
Hey @lgonzalezsa - I think you are on the right track. There is another mention in the docs here: https://mystmd.org/guide/cross-references#targeting-cells - but I don't think it adds much to the other page you found. All you should need is:

- notebook and markdown files in the same directory or a subdirectory, relative to your `myst.yml` file (created with `myst init`)
- notebook cell with the first line as `#| label: my-label`
- reference to this cell in the markdown file from a figure
    ```
    :::{figure} #my-label
    :::
    ```
    
A rich example that includes this feature can be found here: https://github.com/Notebooks-Now/submission-myst-full

In this example there is a labelled cell in `notebooks/visualization-figure-creation-seaborn.ipynb` and a reference to that label in the "results" section of `article.md`.

Hopefully that gets you a little further? Please let me know how it goes and if you run into more trouble!

**Comment by lgonzalezsa** [comment]
Thank you very much for your response and pointers.
I cloned the repository [submission-myst-full](https://github.com/Notebooks-Now/submission-myst-full) and instantiate it locally in my laptop using podman container.

First thing I noticed was during the `myst build --all`, I receive a "Cannot find image" for the notebooks, however, I can see the images if I open the notebook. See the following image, the article does not render the image but the cell output exists in the notebook.

![image](https://github.com/executablebooks/meta/assets/704686/c9964e6a-9036-45b6-8063-abd7a916d9b3)

Now, I am experimenting with the [embed directive](https://mystmd.org/guide/reuse-jupyter-outputs#embed-directive) because i need to use the output of a cell but not as a figure but got an error, see the video attached.


https://github.com/executablebooks/meta/assets/704686/569a7f32-904d-45d8-b59f-6f49eff5fd49

**Reply by rowanc1** [reply]
I have identified the problem with the theme and will have a fix incoming soon!

The PR for the fix is here:
https://github.com/executablebooks/myst-theme/pull/202

I will deploy shortly.

The other error seems to be with the PDF build, @fwkoch any ideas there?
![image](https://github.com/executablebooks/meta/assets/913249/bc0728c8-da5a-4212-8db6-2799d9ceb080)
These seem to be for LaTeX export only, and are likely spurious as we don't need the images from most of the notebook if we are only building the PDF of the markdown. Will have to find a way to mute/remove those, but should be safe to ignore those errors for now.

**Reply by lgonzalezsa** [reply]
Thanks @rowanc1, what do you recommend me to do, in order to test the changes you made?

**Reply by rowanc1** [reply]
Ah, meant to follow up:

```
myst clean --templates
myst start
```

That should delete your `_build/templates` folder and then reinstall the new version!

**Reply by lgonzalezsa** [reply]
Awesome now the embed works!
Thank you for the help!
Any idea why the figure does not render? 
`{figure} #eq-timeline`

![image](https://github.com/executablebooks/meta/assets/704686/4761f38c-261d-4bcc-94cd-4d13951d92b9)

**Reply by rowanc1** [reply]
I haven't been able to reproduce that locally. Are there any warnings about the figure? The label target still exists in the notebook on your copy?

**Reply by lgonzalezsa** [reply]
Interesting, as mentioned in my second post, I cloned the repo and with the fresh copy did not render the image but I can see it in the notebook. I will try without vscode and use Jupyter Lab instead.
![image](https://github.com/executablebooks/meta/assets/704686/a9aca1e5-336a-43c5-ae9f-6be8d3371f0f)

**Reply by lgonzalezsa** [reply]
No difference with Jupyter Lab vs vscode
![image](https://github.com/executablebooks/meta/assets/704686/4aa8b6f8-222a-4005-9437-de442654d4a4)

**Reply by lgonzalezsa** [reply]
Another finding.
Using vscode with the ms-python.black-formatter extension is formatting the frontmatter in my cells from `#| label: label-test` to `# | label: label-test`.
In other words, the black formatter is adding a space between the hash symbol `#` and the pipe symbol `|` which myst seem to skip. Interesting fact, Quarto can handle both formats.

An additional question for the narrative in a separate markdown file. 
I like to create Python cells with IPython display to expose markdown using f strings `f"{variable}"` to write my conclusions calling specific values in Python variables instead of hardcoding in the markdown.

Any tip on how can I achieve the same with Myst?

I used the embed but seems I cannot blend the variable transparently within the paragraph, or could I?
![image](https://github.com/executablebooks/meta/assets/704686/34f03758-a301-4718-b939-e96b32ce11c4)

I would like to write a paragraph in the markdown file invoking a variable or label? from the notebook, here example of what I do today in the same computation notebook:
`f"{count_of_units_this_month} units tested this month."`

Last but not least, I tested the figure directive, and works fine in my repository :100: rendering the figure properly.

**Reply by lgonzalezsa** [reply]
With regard to including variables in my markdown standalone article from another notebook, I continued searching for options and found the `myst_nb.glue()` [https://jupyterbook.org/en/stable/content/executable/output-insert.html#glue-gluing](https://jupyterbook.org/en/stable/content/executable/output-insert.html#glue-gluing)

Now failing due to `unknown directive: glue`

Am I in the right track?

![image](https://github.com/executablebooks/meta/assets/704686/a41450b8-90fb-4e0e-aa27-beaca31bf304)

**Reply by rowanc1** [reply]
>  I tested the figure directive, and works fine in my repository 💯 rendering the figure properly

Wonderful! @fwkoch also tracked down the errors that you saw and has fixed them in https://github.com/executablebooks/mystmd/pull/540 we will do a new release of mystmd later today. Sorry about that, made some pretty large changes recently and missed this path.

If you have a notebook, you can use the `{eval}` directive, currently requires you to use the `jupyterlab-myst` extension. 
See docs on inline expressions:
https://mystmd.org/guide/quickstart-jupyter-lab-myst#inline-expressions

We don't yet support glue in mystmd (some other things we don't support yet are [here](https://github.com/executablebooks/mystmd/issues/189) if you are looking at various docs!).
Because we are not yet evaluating the notebooks, there isn't a way to use `glue`, which is in memory only (AFAIK).

If you evaluate using `jupyterlab-myst`, and then embed the cell, I think it should work -- let me get back to you on that. We also do not yet have a way to expose executed variables outside of a notebook, I would love to have an issue on this to track that work and any initial thoughts you might have about how it would work!!

We will fix up the `# | label:` parsing to support the extra space, will aim to do that shortly as well:
https://github.com/executablebooks/mystmd/issues/541

**Reply by rowanc1** [reply]
Lots of the fixes from this thread are now available in mystmd v1.1.10! As well as a whole lot of improvements to the affiliations and authors by @fwkoch!!

**Reply by lgonzalezsa** [reply]
Awesome guys! way fast!

I was testing with IPython and Markdown to embed a cell output for what I need and seems reasonable alternative to include placeholder to fill with the computation notebook.

```Python
#| label: test-embed
from IPython.display import HTML, Markdown, display
df_ign['DateTime'] = pd.to_datetime(df_ign['Date'] + ' ' + df_ign['Time'])
display(Markdown(f"testing {df_ign['DateTime'][0]} with Markdown"))
```

![image](https://github.com/executablebooks/meta/assets/704686/49c44de6-ab28-4110-afe1-cd06ed76f279)

**Reply by rowanc1** [reply]
Makes me think that we should have an embed *role*, perhaps, as well as a directive, which would be styled differently. As well, it would be good to support markdown as a type a bit easier!

One of the other ideas that I have been kicking around with @stevejpurves is to have a json display output somewhere in your notebook, and then being able to refer to those values in your document.

```md
My study had {var}`#notebook-cell.participants` participants with a p-value of {var}`#notebook-cell.p_value`.
```

Not quite sure about the syntax, regardless would love to have some input! The benefit of this is that it encourages you to expose data from a computation/notebook in a structured way. There could also be a mini library in python that encodes an output as well if we want, maybe it just gets us back to glue? I will open an issue on this and tag you for more discussion.
